sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/PruneShuffle.scala

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.exchange
+
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.SparkPlan
+
+/**
+ * Removes unnecessary shuffles. A shuffle can be introduced by [[Rule]]s for
+ * [[SparkPlan]]s, such as [[EnsureRequirements]] and then, its immediate child of
+ * another shuffle should be unnecessary.
+ */
+case class PruneShuffle() extends Rule[SparkPlan] {
+
+  override def apply(plan: SparkPlan): SparkPlan = plan.transform {
+    case op @ ShuffleExchangeExec(_, child: ShuffleExchangeExec, _) =>
+      op.withNewChildren(Seq(pruneShuffle(child)))
+    case other => other
+  }
+
+  private def pruneShuffle(plan: SparkPlan): SparkPlan = {
+    plan match {
+      case shuffle: ShuffleExchangeExec =>
+        pruneShuffle(shuffle.child)
+      case other => other
+    }
+  }
+}
```

Comment:
I'd propose a more concise way:

```suggestion
case class PruneShuffle() extends Rule[SparkPlan] {

  override def apply(plan: SparkPlan): SparkPlan = plan transformUp {
    case op @ ShuffleExchangeExec(_, ShuffleExchangeExec(_, grandchild, _), _) =>
      op.withNewChildren(grandchild :: Nil)
  }
}
```

Reply:
Hmm, in the current implementation, at most two `ShuffleExchangeExec` can be consecutive. But if it is more consecutive in the future, `transformUp` is not efficient.

Reply:
TBH performance wise we are talking about millisecond level optimization here. I would value readability over micro optimizations.

Reply:
Similar to what is discussed [here](https://github.com/apache/spark/pull/21072#discussion_r183392394), I'd like to avoid unnecessary transformation.